### PR TITLE
Add cudilu

### PR DESCRIPTION
**Pull request by multitalentloes**
Draft version of GPU DILU preconditioner
This DILU implementation uses graph coloring to parallelize the apply and update step of the preconditioner. Supports blocksizes up to 3, and also a float version for benchmarking.

### CODE REVIEW

**Review by kjetilly — COMMENTED** [review]
Generally well done, there are some small beautification fixes.

In general I'd recommend splitting up the cusparse_matrix_operations-file a bit to one only containing DILU-specific stuff.

File: opm/simulators/linalg/cuistl/CuDILU.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    }
+    return res;
+}
+
+std::vector<int>
+createNaturalToReordered(Opm::SparseTable<size_t> levelSets)
+{
+    auto res = std::vector<int>(levelSets.dataSize());
+    int globCnt = 0;
+    for (int i = 0; i < levelSets.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/CuDILU.cpp:66]: Do you see a warning on this particular line? Dont sparseTables return size as an int?

File: opm/simulators/linalg/cuistl/CuDILU.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+// TODO: I think sending size is excessive
+std::vector<int>
+createReorderedToNatural(int size, Opm::SparseTable<size_t> levelSets)
+{
+    auto res = std::vector<int>(size);
+    int globCnt = 0;
+    for (int i = 0; i < levelSets.size(); i++) {
+        for (size_t j = 0; j < levelSets[i].size(); j++) {
+            res[globCnt++] = (int)levelSets[i][j];
```
[the remark this answers is not in the document]
Comment (kjetilly) [reply opm/simulators/linalg/cuistl/CuDILU.cpp]: This was more to guard against wrong use from outside of this function, ie enforce the invariant you have established. It is not strictly needed, but if someone is to change this class further down the line, they might not see the connection to this function.

File: opm/simulators/linalg/cuistl/CuDILU.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kjetilly) [reply opm/simulators/linalg/cuistl/CuDILU.cpp:63]: My mistake, I worked on auto pilot and just assumed size() would return a size_t here, sorry!

File: opm/simulators/linalg/cuistl/detail/cusparse_matrix_operations.cu [hunk 3]
```diff
@@ -21,44 +21,141 @@
...
+        }
+    }
 
-            vecBlock[0] = 1.0 / (diagBlock[0]);
+    // dst -= A*B*C
+    template <class T, int blocksize>
+    __device__ __forceinline__ void mmx2Subtraction(T* A, T* B, T* C, T* dst)
+    {
+
+        T tmp[blocksize * blocksize] = {0};
```
[the remark this answers is not in the document]
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/detail/cusparse_matrix_operations.cu:134]: whats the advantage?

**Comment by multitalentloes** [comment]
I have now gone through the feedback and made the draft PR a regular PR. If this is given the green light I will squash everything into one commit and hopefully it will be merged. As of the only thing missing is to clear up the use of C-style arrays vs std::array for local variables in some of the kernels.

**Comment by kjetilly** [comment]
Jenkins build this please

**Comment by multitalentloes** [comment]
Rebased to simplify history.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
All green, merging. Thanks for the effort!

**Pull request merged**